src/System.Net.Http/src/System/Net/Http/Managed/AuthenticateAndRedirectHandler.cs

```diff
@@ -0,0 +1,223 @@
+                if (location == null)
+                {
+                    // No location header. Nothing to redirect to.
+                    break;
+                }
+
+                if (!location.IsAbsoluteUri)
+                {
+                    location = new Uri(request.RequestUri, location);
+                }
+
+                // Disallow automatic redirection from secure to non-secure schemes
+                bool allowed =
+                    (HttpUtilities.IsSupportedNonSecureScheme(request.RequestUri.Scheme) && HttpUtilities.IsSupportedScheme(location.Scheme)) ||
+                    (HttpUtilities.IsSupportedSecureScheme(request.RequestUri.Scheme) && HttpUtilities.IsSupportedSecureScheme(location.Scheme));
+                if (!allowed)
+                {
+                    break;
+                }
+
+                redirectCount++;
+                if (redirectCount > _maxAutomaticRedirections)
+                {
+                    throw new HttpRequestException(SR.net_http_max_redirects);
+                }
+
+                // Set up for the automatic redirect
+                request.RequestUri = location;
+
+                if (RequestRequiresForceGet(response.StatusCode, request.Method))
+                {
+                    request.Method = HttpMethod.Get;
+                    request.Content = null;
+                }
+
+                // Do the redirect.
+                response.Dispose();
+            }
+
+            return response;
+        }
+
+        private bool RequestNeedsRedirect(HttpResponseMessage response)
+        {
+            // Return if redirect not allowed.
+            if (!_allowRedirect)
+                return false;
+
+            bool needRedirect = false;
+            switch (response.StatusCode)
+            {
+                case HttpStatusCode.Moved:
+                case HttpStatusCode.Found:
+                case HttpStatusCode.SeeOther:
+                case HttpStatusCode.TemporaryRedirect:
+                    needRedirect = true;
+                    break;
+
+                case HttpStatusCode.MultipleChoices:
+                    needRedirect = response.Headers.Location != null; // Don't redirect if no Location specified
+                    break;
+            }
+
+            return needRedirect;
+        }
+
+        private static bool RequestRequiresForceGet(HttpStatusCode statusCode, HttpMethod requestMethod)
+        {
+            if (statusCode == HttpStatusCode.Moved ||
+                statusCode == HttpStatusCode.Found ||
+                statusCode == HttpStatusCode.SeeOther ||
+                statusCode == HttpStatusCode.MultipleChoices)
+            {
+                return requestMethod == HttpMethod.Post;
+            }
+
+            return false;
+        }
+
+        private async Task<HttpResponseMessage> SendRequestAsync(HttpRequestMessage request, bool useCredentialCache, CancellationToken cancellationToken)
+        {
```

Comment:
FWIW, separating this out of the SendAsync method negates the perf benefits that come from combining the auth and redirect handlers.  It's fine for the purposes of this PR, as reducing allocation wasn't the primarily goal and this shouldn't make it worse, but by having these two separate send methods, one of which calls the other that will invariably complete asynchronously, both async methods are going to allocate their state, resulting in two allocations instead of one.